third_party/amd/lib/TritonAMDGPUTransforms/RescheduleOps.cpp

```diff
@@ -1469,76 +1866,195 @@ struct SchedHeuristicLocalLoadAdjacentDotOps
+      if (!addedDotLds)
+        maybeAddSchedBarrier(it, nodeCategoryLocalLoad, nodeCategoryLocalLoad,
+                             "LocalLoadOrder", schedBarMaskBlockDsRead);
+      // LocalStore / LocalStore
+      if (!addedDotLds)
+        maybeAddSchedBarrier(it, nodeCategoryLocalStore, nodeCategoryLocalStore,
+                             "LocalStoreOrder", schedBarMaskBlockDsWrite);
+
+      // GlobalLoad / GlobalLoad
+      if (!addedDotGlobal)
+        maybeAddSchedBarrier(it, nodeCategoryGlobal, nodeCategoryGlobal,
+                             "GlobalLoadOrder", schedBarMaskBlockGlobal);
+    }
+    LDBG("insertSchedBarriers() - DONE");
+  }
 
-  StringRef name() { return "OriginalOrder"; }
+  SchedDag &dag;
+  OpBuilder &builder;
 };
 
 /******************************************************************************
-  SchedManager
-  - Tracks dependencies.
+  SchedManager is top-level struct which manipulates the SchedDag:
+  - Add deps and priorities to dag.
   - Prepared dag for scheduling.
-  - Runs scheduler.
+  - Runs scheduling heuristic on the dag.
 ******************************************************************************/
 struct SchedManager {
-  SchedManager(Block *block) : dag(block), rescheduleId(0) {}
+  SchedManager(Block *block, OpBuilder &builder)
+      : dag(block), rescheduleId(0), builder(builder) {}
 
   // Calculate new deps based on op order and previously determined deps.
   // Insert new deps into dep map and apply them to dat.
   void addDeps(std::unique_ptr<DependencyCalculator> depCalc) {
     depCalc->addDepsToDag(&dag);
   }
 
+  // Calculate new node priorities based on dag.
+  void addPriorities(std::unique_ptr<PriorityCalculator> prioCalc) {
+    prioCalc->addPrioritiesToDag(&dag);
+  }
+
+  // Reschedule the dag.
   template <SchedDirection Direction>
   void reschedule(SchedHeuristic<Direction> *heuristic) {
     LDBG("SchedManager::reschedule("
          << rescheduleId << "), Direction="
          << ((Direction == SchedDirection::TopDown) ? "TopDown" : "BottomUp")
-         << ", Heuristic=" << heuristic->name());
-    // Reset deps right before rescheduling b/c analysis passes
-    // may have altered them.
-    LDBG("dag.resetDeps() before rescheduling");
+         << ", Heuristic=" << heuristic->getName());
+    const bool printDetails = false; // rescheduleId >= 1;
+    // Reset deps right before rescheduling b/c analysis passes may have altered
+    // them.
     dag.resetDeps();
-    LDBG("NodeList before reschedule(" << rescheduleId << ")");
-    LLVM_DEBUG(dag.dumpNodes(llvm::dbgs()));
-    LDBG("SchedDag before reschedule(" << rescheduleId << ")");
-    LLVM_DEBUG(dag.dumpDotFormat(llvm::dbgs()));
-
+    if (printDetails) {
+      LDBG("SchedDag before reschedule(" << rescheduleId << ")");
+      LLVM_DEBUG(dag.dumpDotFormat(llvm::dbgs()));
+    }
     // Node readiness is based on direction.
     dag.initReadyNodes<Direction>();
+    heuristic->begin();
```

Comment:
Maybe instead of 

```cpp
heuristic->begin();
...
heuristic->end();
```

 it is better to use 

```cpp
heuristic->start();
...
heuristic->finish();
```
I thought that `heuristic` was a kind of an `stl` container when I was reading the code the first time

Reply:
Done.